diff --git a/tex.w b/tex.w
index 049968c..c645573 100644
--- a/tex.w
+++ b/tex.w
@@ -10296,7 +10296,7 @@ and |cur_ext|.
 
 @d pack_cur_name	pack_file_name(cur_name, cur_area, cur_ext)
 
-@p void pack_job_name(str_number @!s) /*|s==@[@<|".log"|@>@]|, |@[@<|".dvi"|@>@]|, or
+@p void pack_job_name(str_number @!s) /*|s==@[@<|".log"|@>@]|, |@[@<|".dvi"|@>@]|, |@[@<|".pdf"|@>@]|, or
   |format_extension|*/ 
 {@+cur_area=empty_string;cur_ext=s;
 cur_name=job_name;pack_cur_name;
@@ -10345,18 +10345,22 @@ ship out a box of stuff, we shall use |@<ensure dvi open@>|.
 @<ensure dvi open@>=
 if (output_file_name==0) 
   {@+if (job_name==0) open_log_file();
-  pack_job_name(@[@<|".dvi"|@>@]);
-  while (!b_open_out(&dvi_file)) 
-    prompt_file_name("file name for output",@[@<|".dvi"|@>@]);
+  pack_job_name(output_file_extension);
+  while (!dvi_open_out(&dvi_file)) 
+    prompt_file_name("file name for output",output_file_extension);
   output_file_name=b_make_name_string(&dvi_file);
   } 
 
 @ @<Glob...@>=
+str_number @!output_file_extension;
+
 byte_file @!dvi_file; /*the device-independent output goes here*/ 
 str_number @!output_file_name; /*full name of the output file*/ 
 str_number @!log_name; /*full name of the log file*/ 
 
 @ @<Initialize the output...@>=output_file_name=0;
+if (getenv("engine")) output_file_extension=@<|".dvi"|@>;
+else output_file_extension=@<|".pdf"|@>;
 
 @ The |open_log_file| routine is used to open the transcript file and to help
 it catch up to what has previously been printed on the terminal.
@@ -12775,6 +12779,7 @@ dvi_four(last_bop);last_bop=page_loc;
 cur_v=height(p)+v_offset;temp_ptr=p;
 if (type(p)==vlist_node) vlist_out();@+else hlist_out();
 dvi_out(eop);incr(total_pages);cur_s=-1;
+if (!getenv("engine")) fflush(dvi_file.f);
 done: 
 
 @ Sometimes the user will generate a huge page because other error messages
@@ -12832,12 +12837,25 @@ else{@+dvi_out(post); /*beginning of the postamble*/
     {@+dvi_out(223);decr(k);
     } 
   @<Empty the last bytes out of |dvi_buf|@>;
-  print_nl("Output written on ");slow_print(output_file_name);
+  
+  assert(dvi_close(&dvi_file));
+    print_nl("Output written on "); slow_print(output_file_name);
 @.Output written on x@>
-  print_str(" (");print_int(total_pages);print_str(" page");
-  if (total_pages!=1) print_char('s');
-  print_str(", ");print_int(dvi_offset+dvi_ptr);print_str(" bytes).");
-  b_close(&dvi_file);
+    print_str(" ("); print_int(total_pages); print_str(" page");
+    if (total_pages!=1) print_char('s');
+    if (getenv("engine")) {
+      print_str(", "); print_int(dvi_offset+dvi_ptr); print_str(" bytes).");
+    }
+    else print_str(").");
+  
+  
+    
+
+
+
+ 
+ 
+
   } 
 
 @ @<Output the font definitions...@>=
@@ -26285,6 +26303,9 @@ itself will get a new section number.
 @ 
 @d str_667 "ext4"
 @<|"ext4"|@>=@+667
+@ 
+@d str_668 ".pdf"
+@<|".pdf"|@>=@+668
 
 @ All the above strings together make up the string pool.
 @<|str_pool| initialization@>=
@@ -26340,7 +26361,7 @@ str_632 str_633 str_634 str_635 str_636 str_637 str_638 str_639@/
 str_640 str_641 str_642 str_643 str_644 str_645 str_646 str_647@/
 str_648 str_649 str_650 str_651 str_652 str_653 str_654 str_655@/
 str_656 str_657 str_658 str_659 str_660 str_661 str_662 str_663@/
-str_664 str_665 str_666 str_667 
+str_664 str_665 str_666 str_667 str_668
 
 @ @<|str_start| initialization@>=
 str_start_0_255
@@ -26447,7 +26468,7 @@ str_start_652, str_start_653, str_start_654, str_start_655,
 str_start_656, str_start_657, str_start_658, str_start_659,
 str_start_660, str_start_661, str_start_662, str_start_663,
 str_start_664, str_start_665, str_start_666, str_start_667,
-str_start_668
+str_start_668, str_start_669
 
 @ We still need to define the start locations of the strings.
 @<prepare for string pool initialization@>=
@@ -26865,11 +26886,12 @@ str_start_665=str_start_664+sizeof(str_664)-1,@/
 str_start_666=str_start_665+sizeof(str_665)-1,@/
 str_start_667=str_start_666+sizeof(str_666)-1,@/
 str_start_668=str_start_667+sizeof(str_667)-1,@/
+str_start_669=str_start_668+sizeof(str_668)-1,@/
 str_start_end } str_starts;
 
-@ @<|pool_ptr| initialization@>= str_start_668
+@ @<|pool_ptr| initialization@>= str_start_669
 
-@ @<|str_ptr| initialization@>= 668
+@ @<|str_ptr| initialization@>= 669
 
 @* Index.
 Here is where you can find all uses of each identifier in the program,
