diff --git a/ctangle.w b/ctangle.w
index 1308002..cb30fef 100644
--- a/ctangle.w
+++ b/ctangle.w
@@ -106,19 +106,19 @@ char **av;
 @ The following parameters were sufficient in the original \.{TANGLE} to
 handle \TEX/,
 so they should be sufficient for most applications of \.{CTANGLE}.
-If you change |max_bytes|, |max_names|, or |hash_size| you should also
+If you change |max_bytes|, |max_names|, |buf_size| or |hash_size| you should also
 change them in the file |"common.w"|.
 
-@d max_bytes 90000 /* the number of bytes in identifiers,
+@d max_bytes 150000 /* the number of bytes in identifiers,
   index entries, and section names; used in |"common.w"| */
-@d max_toks 270000 /* number of bytes in compressed \CEE/ code */
-@d max_names 4000 /* number of identifiers, strings, section names;
+@d max_toks 300000 /* number of bytes in compressed \CEE/ code */
+@d max_names 6000 /* number of identifiers, strings, section names;
   must be less than 10240; used in |"common.w"| */
-@d max_texts 2500 /* number of replacement texts, must be less than 10240 */
+@d max_texts 10200 /* number of replacement texts, must be less than 10240 */
 @d hash_size 353 /* should be prime; used in |"common.w"| */
 @d longest_name 10000 /* section names and strings shouldn't be longer than this */
-@d stack_size 50 /* number of simultaneous levels of macro expansion */
-@d buf_size 100 /* for \.{CWEAVE} and \.{CTANGLE} */
+@d stack_size 4000 /* number of simultaneous levels of macro expansion */
+@d buf_size 5000 /* for \.{CWEAVE} and \.{CTANGLE} */
 
 @ The next few sections contain stuff from the file |"common.w"| that must
 be included in both |"ctangle.w"| and |"cweave.w"|. It appears in
