--- tex.w	2020-03-16 15:22:35.103729927 +0700
+++ trip/triptex.w	2020-03-16 15:58:54.971212937 +0700
@@ -401 +401 @@
-enum {@+@!mem_max=30000@+};
+enum {@+@!mem_max=3000@+};
@@ -404 +404 @@
-enum {@+@!mem_min=0@+}; /*smallest index in \TeX's internal |mem| array;
+enum {@+@!mem_min=1@+}; /*smallest index in \TeX's internal |mem| array;
@@ -410,2 +410,2 @@
-enum {@+@!error_line=72@+}; /*width of context lines on terminal error messages*/
-enum {@+@!half_error_line=42@+}; /*width of first lines of contexts in terminal
+enum {@+@!error_line=64@+}; /*width of context lines on terminal error messages*/ 
+enum {@+@!half_error_line=32@+}; /*width of first lines of contexts in terminal
@@ -413 +413 @@
-enum {@+@!max_print_line=79@+}; /*width of longest text lines output; should be at least 60*/
+enum {@+@!max_print_line=72@+}; /*width of longest text lines output; should be at least 60*/ 
@@ -452 +452 @@
-@d mem_bot	0 /*smallest index in the |mem| array dumped by \.{INITEX};
+@d mem_bot	1 /*smallest index in the |mem| array dumped by \.{INITEX};
@@ -454 +454 @@
-@d mem_top	30000 /*largest index in the |mem| array dumped by \.{INITEX};
+@d mem_top	3000 /*largest index in the |mem| array dumped by \.{INITEX};
