--- ../web2w.w	2019-02-22 22:14:57.000000000 +0700
+++ web2w.w	2020-02-18 15:55:32.819147020 +0700
@@ -2624,7 +2624,82 @@
 
 This condition is taken from \.{tex.web}:
 @<Character |k| cannot be printed@>=
-  (k < ' ')||(k > '~')
+  (k < ' ')||(k == 127)||!xchr[k]
+
+@ @<global variables@>=
+wchar_t xchr[256];
+
+@ @<process the command line@>=
+  for (int i = 128; i < 256; i++)
+    xchr[i] = 0;
+  xchr[0x80] = L'А';
+  xchr[0xa0] = L'а';
+  xchr[0x81] = L'Б';
+  xchr[0xa1] = L'б';
+  xchr[0x82] = L'В';
+  xchr[0xa2] = L'в';
+  xchr[0x83] = L'Г';
+  xchr[0xa3] = L'г';
+  xchr[0x84] = L'Д';
+  xchr[0xa4] = L'д';
+  xchr[0x85] = L'Е';
+  xchr[0xa5] = L'е';
+  xchr[0xf0] = L'Ё';
+  xchr[0xf1] = L'ё';
+  xchr[0x86] = L'Ж';
+  xchr[0xa6] = L'ж';
+  xchr[0x87] = L'З';
+  xchr[0xa7] = L'з';
+  xchr[0x88] = L'И';
+  xchr[0xa8] = L'и';
+  xchr[0x89] = L'Й';
+  xchr[0xa9] = L'й';
+  xchr[0x8a] = L'К';
+  xchr[0xaa] = L'к';
+  xchr[0x8b] = L'Л';
+  xchr[0xab] = L'л';
+  xchr[0x8c] = L'М';
+  xchr[0xac] = L'м';
+  xchr[0x8d] = L'Н';
+  xchr[0xad] = L'н';
+  xchr[0x8e] = L'О';
+  xchr[0xae] = L'о';
+  xchr[0x8f] = L'П';
+  xchr[0xaf] = L'п';
+  xchr[0x90] = L'Р';
+  xchr[0xe0] = L'р';
+  xchr[0x91] = L'С';
+  xchr[0xe1] = L'с';
+  xchr[0x92] = L'Т';
+  xchr[0xe2] = L'т';
+  xchr[0x93] = L'У';
+  xchr[0xe3] = L'у';
+  xchr[0x94] = L'Ф';
+  xchr[0xe4] = L'ф';
+  xchr[0x95] = L'Х';
+  xchr[0xe5] = L'х';
+  xchr[0x96] = L'Ц';
+  xchr[0xe6] = L'ц';
+  xchr[0x97] = L'Ч';
+  xchr[0xe7] = L'ч';
+  xchr[0x98] = L'Ш';
+  xchr[0xe8] = L'ш';
+  xchr[0x99] = L'Щ';
+  xchr[0xe9] = L'щ';
+  xchr[0x9a] = L'Ъ';
+  xchr[0xea] = L'ъ';
+  xchr[0x9b] = L'Ы';
+  xchr[0xeb] = L'ы';
+  xchr[0x9c] = L'Ь';
+  xchr[0xec] = L'ь';
+  xchr[0x9d] = L'Э';
+  xchr[0xed] = L'э';
+  xchr[0x9e] = L'Ю';
+  xchr[0xee] = L'ю';
+  xchr[0x9f] = L'Я';
+  xchr[0xef] = L'я';
+  xchr[0xfc] = L'№';
+
 @
 
 
